File: scripts/reference-solutions/meeting-rooms-ii.cpp
#include <algorithm>
#include <functional>
#include <iostream>
#include <queue>
#include <utility>
#include <vector>

using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  int n;
  if (!(cin >> n)) {
    cout << "0\n";
    return 0;
  }

  vector<pair<long long, long long>> intervals;
  intervals.reserve(n);
  for (int i = 0; i < n; ++i) {
    long long start, end;
    cin >> start >> end;
    if (start < end) {
      intervals.push_back({start, end});
    }
  }

  sort(intervals.begin(), intervals.end());
  priority_queue<long long, vector<long long>, greater<long long>> rooms;
  int answer = 0;

  for (const auto& [start, end] : intervals) {
    while (!rooms.empty() && rooms.top() <= start) {
      rooms.pop();
    }
    rooms.push(end);
    answer = max(answer, static_cast<int>(rooms.size()));
  }

  cout << answer << '\n';
  return 0;
}

File: scripts/reference-solutions/meeting-rooms-ii.py
import heapq
import sys

data = sys.stdin.read().strip().split()
if not data:
    print(0)
    sys.exit(0)

n = int(data[0])
intervals = []
index = 1
for _ in range(n):
    start = int(data[index])
    end = int(data[index + 1])
    index += 2
    if start < end:
        intervals.append((start, end))

intervals.sort()
rooms = []
answer = 0

for start, end in intervals:
    while rooms and rooms[0] <= start:
        heapq.heappop(rooms)
    heapq.heappush(rooms, end)
    answer = max(answer, len(rooms))

print(answer)
